Add intrinsics for the FEAT_SVE2p2/FEAT_SME2p2 features introduced by the 2024 dpISA

amilendra opened the pull request:
---
name: Pull request
about: Technical issues, document format problems, bugs in scripts or feature proposal.

---

<!-- SPDX-FileCopyrightText: Copyright 2021-2022 Arm Limited and/or its affiliates <open-source-office@arm.com> -->
<!-- SPDX-License-Identifier: Apache-2.0 -->

**Thank you for submitting a pull request!**

If this PR is about a bugfix:

Please use the bugfix label and make sure to go through the checklist below.

If this PR is about a proposal:

We are looking forward to evaluate your proposal, and if possible to
make it part of the Arm C Language Extension (ACLE) specifications.

We would like to encourage you reading through the [contribution
guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md), in particular the section on [submitting
a proposal](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#proposals-for-new-content).

Please use the proposal label.

As for any pull request, please make sure to go through the below
checklist.

Checklist: (mark with ``X`` those which apply)

* [X] If an issue reporting the bug exists, I have mentioned it in the
      PR (do not bother creating the issue if all you want to do is
      fixing the bug yourself).
* [X] I have added/updated the `SPDX-FileCopyrightText` lines on top
      of any file I have edited. Format is `SPDX-FileCopyrightText:
      Copyright {year} {entity or name} <{contact informations}>`
      (Please update existing copyright lines if applicable. You can
      specify year ranges with hyphen , as in `2017-2019`, and use
      commas to separate gaps, as in `2018-2020, 2022`).
* [X] I have updated the `Copyright` section of the sources of the
      specification I have edited (this will show up in the text
      rendered in the PDF and other output format supported). The
      format is the same described in the previous item.
* [X] I have run the CI scripts (if applicable, as they might be
      tricky to set up on non-*nix machines). The sequence can be
      found in the [contribution
      guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration). Don't
      worry if you cannot run these scripts on your machine, your
      patch will be automatically checked in the Actions of the pull
      request.
* [X] I have added an item that describes the changes I have
      introduced in this PR in the section **Changes for next
      release** of the section **Change Control**/**Document history**
      of the document. Create **Changes for next release** if it does
      not exist. Notice that changes that are not modifying the
      content and rendering of the specifications (both HTML and PDF)
      do not need to be listed.
* [X] When modifying content and/or its rendering, I have checked the
      correctness of the result in the PDF output (please refer to the
      instructions on [how to build the PDFs
      locally](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration)).
* [X] The variable `draftversion` is set to `true` in the YAML header
      of the sources of the specifications I have modified.
* [ ] Please *DO NOT* add my GitHub profile to the list of contributors
      in the [README](https://github.com/ARM-software/acle/blob/main/README.md#contributors-) page of the project.

amilendra committed:
Add support for FEAT_SVE2p2/FEAT_SME2p2 intrinsics

These instructions are available under features FEAT_SVE2p2 or FEAT_SME2p2.

COMPACT: Copy Active vector elements to lower-numbered elements (Byte/Halfword variants)
EXPAND: Copy lower-numbered vector elements to Active elements (Byte/Halfword/Word/Doubleword variants)
FIRSTP: Scalar index of first true predicate element (predicated) (Byte/Halfword/Word/Doubleword variants)
LASTP: Scalar index of last true predicate element (predicated) (Byte/Halfword/Word/Doubleword variants)
FMUL (multiple and single vector): Multi-vector floating-point multiply by vector
FMUL (multiple vectors): Multi-vector floating-point multiply

AlfieRichardsArm commented on main/acle.md:
```diff
@@ -13549,6 +13559,63 @@ While (resulting in predicate tuple)
+
+These intrinsics can only be called from non-streaming code if
+`__ARM_FEATURE_SVE2p2` is defined. They can only be called from streaming code
+if the appropriate SME feature macro is defined (see previous paragraph).
```
`... appropriate  SME2p2 feature macro is defined...` instead?

AlfieRichardsArm commented on main/acle.md:
````diff
@@ -13549,6 +13559,63 @@ While (resulting in predicate tuple)
+``` c
+  // Variants are available for:
+  // _b16, _b32, _b64
+  svbool_t svfirstp[_b8](svbool_t pg, svbool_t op);
````
I am not well versed on the underlying instructions, but shouldn't they return an integer of some kind?

amilendra replied:
Yes. You are right, the resulting index is stored in a general purpose X register and can hold -1 when a matching element is not found, so I've fixed the return type to be `int64_t`. Thanks

amilendra commented on main/acle.md:
```diff
@@ -13549,6 +13559,63 @@ While (resulting in predicate tuple)
+ [`<arm_sve.h>`](#arm_sve.h) or [`<arm_sme.h>`](#arm_sme.h)
+when `__ARM_FEATURE_SVE2p2` or `__ARM_FEATURE_SME2p2` is defined, respectively.
+
+These intrinsics can only be called from non-streaming code if
```
I've updated the conditions. Please can you have another look?

Lukacma reviewed (approved):
LGTM

amilendra committed:
Add SME2.2/SVE2.2 support status (Alpha) to the content

amilendra commented on main/acle.md:
```diff
@@ -13426,6 +13593,56 @@ While (resulting in predicate tuple)
+
+Scalar index of first/last true predicate element (predicated).
+
+These intrinsics can be called from streaming mode if either of the feature
```
Similar to COMPACT, this is also something I understood from the [CheckSVEEnabled](https://developer.arm.com/documentation/ddi0602/2025-06/Shared-Pseudocode/aarch64-functions-sve?lang=en#impl-aarch64.CheckSVEEnabled.0) pseudocode used by [FIRSTP](https://developer.arm.com/documentation/ddi0602/2025-06/SVE-Instructions/FIRSTP--Scalar-index-of-first-true-predicate-element--predicated--?lang=en)

amilendra committed:
SME/SVE2.2 Remove conditions for being called by streaming/non-streaming code

amilendra committed:
Address review comments

1. Change firstp/lastp parameter names to match the register names: op -> pn
2. split the x2 and x4 variants of the FMUL intrinsics

Lukacma commented on main/acle.md:
```diff
@@ -12927,6 +12937,41 @@ Zero ZA vector groups
+extended in the future.
+
+The intrinsics in this section are defined by the header file
+[`<arm_sme.h>`](#arm_sme.h) when `__ARM_FEATURE_SME2p2` is defined.
```
Since these instructions do not operate on ZA array,  I think they should be in arm_sve.h.

Lukacma replied:
After internal discusion this whole system of specifying header files will be overhauled, so no changes necessary here until then

Lukacma committed:
add frint intrinsics

Lukacma committed:
remove empty lines

Lukacma committed:
Add missing zeroing forms of  predicated convert intrinsics

CarolineConcatto commented on main/acle.md:
````diff
@@ -13476,6 +13521,97 @@ While (resulting in predicate tuple)
+```c
+
+// Variant is available for _f64
+svfloat32_t frint32x[_f32]_z(svbool_t pg, svfloat32_t zn);
````
I believe this should be:
s /frint/svrint/g

Lukacma committed:
Fix frint intrinsics and formatting errors

Lukacma commented on main/acle.md:
```diff
@@ -13476,6 +13521,97 @@ While (resulting in predicate tuple)
+// _f32_f64, _bf16_f32
+svfloat16_t	svcvtnt_f16[_f32]_z	(svfloat16_t even, svbool_t pg, svfloat32_t op)
+
+svfloat32_t	svcvtxnt_f32[_f64]_z	(svfloat32_t even, svbool_t pg, svfloat64_t op);
```
Done

artemiy-volkov-arm reviewed (approved):
LGTM

Lukacma committed:
Merge branch 'main' into 2025-acle-sve-sme2p2

vhscampos reviewed (approved):
LGTM. Merging

vhscampos committed:
Merge branch 'main' into 2025-acle-sve-sme2p2

sallyarmneale reviewed (commented):
no changes required.
